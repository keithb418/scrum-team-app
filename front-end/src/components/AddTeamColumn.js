import React from "react";
import { Button } from "react-bootstrap";
import FontAwesome from "react-fontawesome";

const AddTeamColumn = () => {
  return (
    <Button className="add-team-button">
      <span className="content">
        <FontAwesome
          name="plus-circle"
        />
        Add Team Column
      </span>
    </Button>
  );
};

export default AddTeamColumn;
