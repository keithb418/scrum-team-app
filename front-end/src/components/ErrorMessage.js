import React from "react";
import { Alert } from "react-bootstrap";

const ErrorMessage = ({ message }) =>
  message
    ? <div><Alert bsStyle="warning"><strong>Error:</strong> {message}</Alert></div>
    : null;

export default ErrorMessage;