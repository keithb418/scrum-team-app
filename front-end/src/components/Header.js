import React from "react";
import { Row, Col } from "react-bootstrap";

const Header = () => {
  return (
    <header>
      <Row>
        <Col xs={12}>
          <div className="page-header">
            <h1>Scrum Team</h1>
          </div>
        </Col>
      </Row>
    </header>
  );
};

export default Header;
