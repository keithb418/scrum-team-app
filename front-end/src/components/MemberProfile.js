import React, { Component } from "react";
import { connect } from "react-redux";
import { Button, Row, Col, Panel } from "react-bootstrap";
import { fetchProfileData } from "../actions/shared";
import FontAwesome from "react-fontawesome";
import { Link } from "react-router-dom";
import PropTypes from "prop-types";
import { capitalizeFirstLetter } from "../utils/stringHelpers";
import Loading from "./Loading";
import Profile from "./Profile";
import Skills from "./Skills";

export class MemberProfile extends Component {
  constructor () {
    super();
  }
 
  componentDidMount () {
    this.props.fetchProfileData(this.props.match.params.id);
  }
 
  render () {
    const { teamMember, team } = this.props;
    return !teamMember ?
        <Loading /> :
        <Row>
          <Col className="profile-section" md={6} mdOffset={3} xs={12}>
            <Panel.Heading>
              <h3 className="panel-title">{team.name}</h3>
            </Panel.Heading>
            <hr />
            <div className="member-profile-subheading">
              <FontAwesome name="user-circle" className="member-img" />
              <p>{teamMember.name}</p>
            </div>
            <hr />
            <Row>
              <Profile 
                member={teamMember}/>
              <Skills 
                member={teamMember}/>
            </Row>
            <hr />
            <Link to={`/member/${teamMember._id}/edit`}>
              <Button bsStyle="primary">Edit</Button>
            </Link>
            <Link to="/">
              <Button>Back to Dashboard</Button>
            </Link>
          </Col>
        </Row>;
  }
};

MemberProfile.defaultProps = {
  team: {
    name: ""
  },
  isFetching: true
};

MemberProfile.propTypes = {
  teamMember: PropTypes.object,
  team: PropTypes.object
};

const mapStateToProps = ({ teamMembers: { teamMember }, teams: { teams } }) => {
  if(teamMember) {
    const team = teams.find(team => {
      return team._id === teamMember.team;
    });
    return {
      team,
      teamMember
    };
  } else {
    return {
      teamMember
    };
  }  
};

const mapDispatchToProps = (dispatch) => ({
  fetchProfileData: (id) => dispatch(fetchProfileData(id))
});


export default connect(mapStateToProps, mapDispatchToProps)(MemberProfile);