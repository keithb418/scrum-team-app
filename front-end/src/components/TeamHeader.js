import React from "react";
import { Button, Row, Col } from "react-bootstrap";
import { connect } from "react-redux";
import FontAwesome from "react-fontawesome";
import PropTypes from "prop-types";
import { Link } from "react-router-dom";
import { truncateString } from "../util/stringHelpers";

let TeamHeader = ({ teamName, deleteTeam, id, teamMembers }) => {

  const isDisabled = teamMembers.length > 0;

  return (
    <div className="team-header panel-heading">
      <Link to={`team/${id}/member/add`}>
        <Button className="add-team-member-btn">
          <FontAwesome name="user-plus" />
        </Button>
      </Link>
        <h2 className="col-xs-10 panel-title">{truncateString(teamName, 16)}</h2>
      <Link to={`team/${id}/edit`}>
        <Button>
          <FontAwesome name="edit" />
        </Button>
      </Link>
      <Button
        disabled={isDisabled}
        onClick={() => deleteTeam(id)}>
        <FontAwesome name="trash" />
      </Button>
    </div>
  );
};

TeamHeader = connect()(TeamHeader);

TeamHeader.propTypes = {
  teamName: PropTypes.string,
  projectName: PropTypes.string,
  dispatch: PropTypes.func,
  id: PropTypes.string
};

export default TeamHeader;