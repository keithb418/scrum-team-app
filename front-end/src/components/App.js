import React from "react";
import TeamColumnsContainer from "../containers/TeamColumnsContainer";
import AddTeamMember from "./forms/AddTeamMember";
import Header from "./Header";

const App = () => (
  <div>
    <Header />
    <TeamColumnsContainer />
    <AddTeamMember />
  </div>
);

export default App;
