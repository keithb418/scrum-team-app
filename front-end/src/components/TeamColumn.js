import React, { Component } from "react";
import { connect } from "react-redux";
import { handleChangeTeam } from "../actions/teamMembers";
import { handleDeleteTeam } from "../actions/teams";
import TeamMember from "./TeamMember";
import TeamHeader from "./TeamHeader";
import PropTypes from "prop-types";

export class TeamColumn extends Component {
  constructor(props) {
    super(props)
  }

  allowDrop = (e) => {
    e.preventDefault();
  };

  drop = (e) => {
    e.preventDefault();
    const _id = e.dataTransfer.getData("tmId");
    const team = this.props.id;
    this.props.changeTeam(_id, team);
  };

  handleDelete = () => {
    this.props.deleteTeam(this.props.id)
  }

  render() {
    const { id, teamName, teamMembers, project } = this.props;
    return (
      <div className="panel panel-default team-column" onDragOver={this.allowDrop} onDrop={this.drop}>
        <TeamHeader
          teamMembers={teamMembers}
          id={id}
          teamName={teamName} />
        <div className="panel-body team-body" onDrop={this.drop} onDragOver={this.allowDrop} >
          { teamMembers.map(teamMember =>
            <TeamMember
              key={teamMember._id}
              id={teamMember._id}
              name={teamMember.name}
              teamLead={teamMember.teamLead}
              role={teamMember.role} />
          )}
        </div>
      </div>
    );
  }  
}

TeamColumn.defaultProps = {
  id: "",
  teamName: "",
  teamMembers: []
}

TeamColumn.propTypes = {
  id: PropTypes.string,
  teamMembers: PropTypes.array,
  teamName: PropTypes.string,
};

const mapDispatchToProps = dispatch => ({
  changeTeam: (_id, team) => dispatch(handleChangeTeam(_id, team)),
})

export default connect(undefined, mapDispatchToProps)(TeamColumn);