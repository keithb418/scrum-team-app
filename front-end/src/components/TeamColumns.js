import React, { Component } from 'react'
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { fetchInitialData } from '../actions/shared';
import TeamColumn from "./TeamColumn";
import AddTeamColumn from "./AddTeamColumn";
import Loading from "./Loading";

export class TeamColumns extends Component { 
  constructor(props) {
    super(props)
  }

  componentDidMount() {
    this.props.fetchInitialData();
  }

  render() {
    const { teams } = this.props
    return (
      this.props.isFetching === true ?
      <Loading /> :
      <div className="team-columns">
        {teams.map(team =>
          <TeamColumn
            key={team._id}
            id={team._id}
            teamName={team.name}
            teamMembers={team.teamMembers}
          />
        )}
        <AddTeamColumn />
      </div>
    )
  }
}


TeamColumns.propTypes = {
  teams: PropTypes.array
};

const mapStateToProps = ({teams: { teams }, teamMembers:{ teamMembers }, fetch: { isFetching }}) => {
  let teamsList = teams.map(item => ({ ...item }))
  let teamMembersList = teamMembers.map(item => ({ ...item }))

  teamMembersList
  .map((teamMember) => {
    let team = teamsList.find((team) => team._id === teamMember.team);

    if (team) {
      team.teamMembers = team.teamMembers || [];
      team.teamMembers.push(teamMember);
    }
  })
  .sort((a, b) => {
    if (b.teamLead) {
      return 1;
    }

    if (a.teamLead) {
      return -1;
    }
    return 0;
  });

  return {
    teams: teamsList,
    isFetching
  };
}

const mapDispatchToProps = (dispatch) => ({
  fetchInitialData: () => dispatch(fetchInitialData())
});

export default connect(mapStateToProps, mapDispatchToProps)(TeamColumns);