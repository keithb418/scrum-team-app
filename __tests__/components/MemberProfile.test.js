import React from 'react';
import { shallow } from 'enzyme';
import teamMemberData from '../fixtures/teamMembers';
import teamsData from '../fixtures/teams';
import { MemberProfile } from '../../front-end/src/components/MemberProfile';
let fetchProfileData;

test('should render the MemberProfile component correctly', () => {
    fetchProfileData = jest.fn();
    const path = {
        params: {
            id: teamMemberData[1]._id
        }
    }
    const wrapper = shallow(
    <MemberProfile 
        fetchProfileData={fetchProfileData}
        match={path}
        team={teamsData[1]} 
        teamMember={teamMemberData[1]} 
    />
);
    expect(wrapper).toMatchSnapshot();
    expect(fetchProfileData).toHaveBeenCalledWith(teamMemberData[1]._id);
});

