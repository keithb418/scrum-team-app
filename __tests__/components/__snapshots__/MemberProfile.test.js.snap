// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the MemberProfile component correctly 1`] = `
<Row
  bsClass="row"
  componentClass="div"
>
  <Col
    bsClass="col"
    className="profile-section"
    componentClass="div"
    md={6}
    mdOffset={3}
    xs={12}
  >
    <PanelHeading
      bsClass="panel"
      componentClass="div"
    >
      <h3
        className="panel-title"
      >
        Team 2
      </h3>
    </PanelHeading>
    <hr />
    <div
      className="member-profile-subheading"
    >
      <FontAwesome
        className="member-img"
        name="user-circle"
      />
      <p>
        Dennis
      </p>
    </div>
    <hr />
    <Row
      bsClass="row"
      componentClass="div"
    >
      <Profile
        member={
          Object {
            "_id": "2",
            "email": "dtest@mail.com",
            "experience": "3 years",
            "name": "Dennis",
            "role": "Fullstack Developer",
            "skills": Array [
              "react",
            ],
            "team": "2abc",
            "teamLead": false,
          }
        }
      />
      <Skills
        member={
          Object {
            "_id": "2",
            "email": "dtest@mail.com",
            "experience": "3 years",
            "name": "Dennis",
            "role": "Fullstack Developer",
            "skills": Array [
              "react",
            ],
            "team": "2abc",
            "teamLead": false,
          }
        }
      />
    </Row>
    <hr />
    <Link
      replace={false}
      to="/member/2/edit"
    >
      <Button
        active={false}
        block={false}
        bsClass="btn"
        bsStyle="primary"
        disabled={false}
      >
        Edit
      </Button>
    </Link>
    <Link
      replace={false}
      to="/"
    >
      <Button
        active={false}
        block={false}
        bsClass="btn"
        bsStyle="default"
        disabled={false}
      >
        Back to Dashboard
      </Button>
    </Link>
  </Col>
</Row>
`;
